deepwell/src/services/site_member/service.rs

```diff
@@ -0,0 +1,158 @@
+ * Copyright (C) 2019-2023 Wikijump Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use super::prelude::*;
+use crate::models::site_member::{self, Entity as SiteMember, Model as SiteMemberModel};
+
+#[derive(Debug)]
+pub struct SiteMemberService;
+
+impl SiteMemberService {
+    /// Add a user to a site.
+    pub async fn add(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<Option<SiteMemberModel>> {
+        let txn = ctx.transaction();
+        tide::log::info!(
+            "Adding membership of user with ID {user_id} to site ID {site_id}"
+        );
+
+        // If the user is already a member of the target site, discontinue.
+        if Self::get_optional(ctx, SiteMembership { site_id, user_id })
+            .await?
+            .is_some()
+        {
+            return Ok(None);
+        }
+
+        // TODO: Check for membership qualifications (e.g. bans).
+
+        // Insert new membership.
+        let model = site_member::ActiveModel {
+            user_id: Set(user_id),
+            site_id: Set(site_id),
+            deleted_at: Set(None),
+            ..Default::default()
+        };
+
+        let membership = model.insert(txn).await?;
+
+        Ok(Some(membership))
+    }
+
+    /// Remove a user from a site.
+    pub async fn remove(
+        ctx: &ServiceContext<'_>,
+        SiteMembership { site_id, user_id }: SiteMembership,
+    ) -> Result<SiteMemberModel> {
+        let txn = ctx.transaction();
+        tide::log::info!(
+            "Removing the membership of user ID {user_id} from site ID {site_id}"
+        );
+
+        let model = match Self::get_optional(ctx, SiteMembership { site_id, user_id })
+            .await?
+        {
+            // If membership is found, remove it by setting the leave date.
+            Some(member_model) => {
+                let mut model = member_model.into_active_model();
+                model.deleted_at = Set(Some(now()));
+                model.update(txn).await?
+            }
+
+            // If no membership is found, return BadRequest error.
+            None => {
+                tide::log::error!(
+                    "Could not remove user ID {user_id} from site ID {site_id} as they are not a member."
+                );
+                return Err(Error::BadRequest);
```

Comment:
Couldn't this return `None` for non-members? I was thinking add and removal operations could both be idempotent.

Reply:
Sure, I could change it to that. I was just modelling it after what VoteService did.

Reply:
Hmm. Yeah no I think you're right it should stay like this like for vote service.